DOCS-7626 converted policy to template (#236)
mulesoft/docs-api-manager: modules/ROOT/pages/ip-blacklist.adoc

[width="100%", cols="5,15"]
|===
>s| Policy Name | IP Blacklist

> >s|Summary      | Blocks access from a range of IP addresses to an API endpoint

Reviewer: Does this mean...:
```suggestion
>s|Summary      | Blocks a range of IP addresses from accessing an API endpoint
```

Author: Done! Thanks!

>s|Category | Security
>s|First Mule version available  | 3.8.0
.1+>.^s| Returned Status Codes
| 403 - IP is rejected
|===

The IP Blacklist policy controls access to a configured API endpoint from a range of IP addresses. 

The IP Blacklist policy restricts access to a protected resource when a match is found between a source IP (specified when configuring the policy) and a list of individual IPs or range of IPs.

> Based on your business need, you can specify a DataWeave 2.0 expression for Mule runtime engine 4 (Mule)) or MEL expression for Mule 3.

Reviewer 2: The reason of why could be useful to do that is missing. The expression can be specified to define which would be the source IP to be used when the policy tries to determine if the IP belongs to the restricted IP list defined in the policy or not.

Author: Done! Thanks!

> Based on your business need, you can specify a DataWeave 2.0 expression for Mule runtime engine 4 (Mule)) or MEL expression for Mule 3.

Reviewer: ```suggestion
Based on your business need, you can specify a DataWeave 2.0 expression for Mule runtime engine 4 (Mule) or MEL expression for Mule 3.
```

Author: Done! Great catch! Thanks!

[[specifying_ip_address]]
== Specifying IP Addresses to Blacklist in the Policy

> You can specify the IP addresses to block in the policy using the following ways:

Reviewer: ```suggestion
You can specify the IP addresses to block in the policy in the following ways:
```

Author: Done! Thanks!

> * Blacklist access based on the IP address of the request

Reviewer: ```suggestion
* Blacklist access based on the IP address of the request.
```

Author: Done! Thanks!

> * Blacklist access based on the origin IP address that is determined by resolving a specific DataWeave or MEL expression, such as the https://en.wikipedia.org/wiki/X-Forwarded-For[x-Forwarded-For] header or the `HTTP` header.

Reviewer 2: HTTP header is the generic way to mention a header that is part of a HTTP request or response. It is not a special header. A header is composed by a key and a value. The key can be, for example: `x-Forwarded-For`.

Reviewer: Looks like x-Forwarded-For should be 'x-Forwarded-For` (formatted)?

Author: @nahuel-dallavecchia removed the HTTP part from the sentence. @kbostrom formatting does not work inside []. Done! Thanks!

* Any other origin.

=== Blacklist Access Based on the IP Address of the Request

To blacklist access based on the IP address of the request, you can:

> 

Reviewer: Should the following list items begin with "Block" or "Blacklist" instead of "Define"?

Author: No I think define is fine. I think either define or specify works. Done! Thanks!

* Define a specific IP address by enumerating it in the white space, for example, `192.168.1.1`.
* Define a subset of addresses by identifying a subnet mask. 
+

> For example, `192.168.3.1/30` includes the consolidated range: `192.168.3.0 - 192.168.3.3`.

Reviewer: ```suggestion
For example, `192.168.3.1/30` includes the consolidated range `192.168.3.0 - 192.168.3.3`.
```

Author: Done! Thanks!

* Define a whole range of IP addresses by specifying the relevant octets of the IP address that you want to permit. 
+

> For example, setting `192.168` will include IP addresses from 192.168.0.0 to 192.168.255.255.

Reviewer: "to" or "through" ....255?

Author: Done! Thanks!

=== Blacklist Access Based on the Origin IP Address of the x-Forwarded-For Header

> If the client connects to your API through an HTTP proxy or a load-balancer, you can blacklist the client's specific IP (the IP originating the request) instead of the address that will appear in the request.

Reviewer: ```suggestion
If the client connects to your API through an HTTP proxy or a load balancer, you can blacklist the client's specific IP address (the IP address originating the request) instead of the address that appears in the request.
```

Author: Done! Thanks!

> For example, suppose you want to blacklist `192.168.2.3` and the address of a client connecting through an HTTP proxy is `92.40.1.255`, the client requests appear with this public address using the proxy in such a case. 

Reviewer: ```suggestion
For example, if you want to blacklist `192.168.2.3` and the address of a client connecting through an HTTP proxy is `92.40.1.255`, the client requests appear with the public address using the proxy. 
```

Author: Done! Thanks!

Typically, applications use the `X-Forwarded-For` header to identify the origin IP addresses of a request that was redirected to your endpoint.

> You can use a Mule expression to instruct the service to look for the IP address in the 'x-Forwarded-For' header:

Reviewer 2: Mule expression here means MEL or DataWeave 2.0, right?

Author: Done! Thanks!

[width="100%", cols="5,15"]
|===
>s| Policy Name | IP Blacklist
>s|Summary      | Blocks access from a range of IP addresses to an API endpoint
>s|Category | Security
>s|First Mule version available  | 3.8.0
.1+>.^s| Returned Status Codes
| 403 - IP is rejected
|===

The IP Blacklist policy controls access to a configured API endpoint from a range of IP addresses. You can blacklist access based on the IP address of the request

> or based on the origin IP address of the `x-Forwarded-For` header.

Reviewer 2: The statement is not completely correct. Extracting the IP address from the x-forwarded-for header is ok, but it can be extracted from other HTTP header, or from some completely different place, taking into account that you may specify the DataWeave 2.0 expression (Mule 4) or MEL expression (Mule 3) that better suits your business need.

Author: Done! Thanks!

> //Nauel--will users know what this header is?

Reviewer 2: The `x-forwarded-for` HTTP header is a well known header used by many Internet routers, so the route taken by a message can be audited for example (More info in: https://en.wikipedia.org/wiki/X-Forwarded-For).

Author: Done! Thanks!

== How This Policy Works

> // Nahuel--anything to add here?

Reviewer 2: The policy restricts access to a protected resource when there is some match between a source IP (which is specified when configuring the policy) and a list of individual IPs or range of IPs.

Author: Done! Thanks!

[[specifying_ip_address]]
== Specifying IP Addresses to Blacklist in the Policy

You can specify the IP addresses to block in the policy using the following ways:

* Blacklist access based on the IP address of the request

> * Blacklist access based on the origin IP address of the x-Forwarded-For header

Reviewer 2: Also blacklist access based on the origin IP address that is determined by resolving a specific expression (`x-forwarded-for` is a particular case of this).

Author: Done! Thanks!

=== Blacklist Access Based on the IP Address of the Request

To blacklist access based on the IP address of the request, you can:

* Define a specific IP address by enumerating it in the white space, for example, `192.168.1.1`.
* Define a subset of addresses by identifying a subnet mask. 
+
For example, `192.168.3.1/30` includes the consolidated range: `192.168.3.0 - 192.168.3.3`.
* Define a whole range of IP addresses by specifying the relevant octets of the IP address that you want to permit. 
+
For example, setting `192.168` will include IP addresses from 192.168.0.0 to 192.168.255.255.

=== Blacklist Access Based on the Origin IP Address of the x-Forwarded-For Header

If the client connects to your API through an HTTP proxy or a load-balancer, you can blacklist the client's specific IP (the IP originating the request) instead of the address that will appear in the request.

For example, suppose you want to blacklist `192.168.2.3` and the address of a client connecting through an HTTP proxy is `92.40.1.255`, the client requests appear with this public address.

> //Nahuel--which public address. The previous para is not clear as to what we want to say. 

Reviewer 2: So you try to access a protected resource. You have an IP assigned by your company's router, for example. Let's say it's `192.168.2.3` (-> private IP). But the company has 1 Internet provider and only one IP that can be used to surf the Internet. For example `92.40.1.255` (-> public IP). So you may always have a static IP `192.168.2.3` but the public IP address may change without notice, and you'd like to blacklist the original private IP. In this scenario, all outgoing communications from the company will appear as coming from `92.40.1.255`, no mattering the actual private IP where the message was originated.

Author: Done! Thanks!

#[message.inboundProperties['http.headers']['X-Forwarded-For']]
----

> When you insert the IP address in the IP expression field of the policy parameters, the API Platform is instructed to look at the first IP address that ranks in the concatenated values of the `x-Forwarded-For` header of the request:

Reviewer: API Platform > API platform? Anypoint Platform API?

Author: Anypoint Platform. Done! Thanks!

image::ip-whitelist-ip-expression.png[ip+blacklist+ip_expression]

== Configuring Policy Parameters

When you apply the IP Blacklist policy to your API from the UI, you can configure the following parameters:

[%header%autowidth.spread,cols="a,a,a"]
|===
| Parameter | Description 
|IP expression  |The DataWeave 2.0 or MEL expression to be used for extracting the IP address from this API request. 

> |Blacklist |The list of IP addresses that must be denies access to the API. You may define one IP or an IP range at a time, as many times required. For more information, see <<specifying_ip_address,Specifying IP Addresses to Blacklist in the Policy>>.

Reviewer: ```suggestion
|Blacklist |The list of IP addresses blocked from the API. You can define one IP or an IP range at a time, as many times required. For more information, see <<specifying_ip_address,Specifying IP Addresses to Blacklist in the Policy>>.
```

Author: Done! Thanks!

#[message.inboundProperties['http.headers']['X-Forwarded-For']]
----

When you insert the IP address in the IP expression field of the policy parameters, the API Platform is instructed to look at the first IP address that ranks in the concatenated values of the `x-Forwarded-For` header of the request:

image::ip-whitelist-ip-expression.png[ip+blacklist+ip_expression]

== Configuring Policy Parameters

When you apply the IP Blacklist policy to your API from the UI, you can configure the following parameters:

[%header%autowidth.spread,cols="a,a,a"]
|===
| Parameter | Description 
|IP expression  

> |The Mule expression to be used for extracting the IP address from this API request. 

Reviewer 2: In Mule 4 is a DataWeave 2.0 expression, which is quite similar to MEL in this case, but not the same.

Author: Done! Thanks!

|Blacklist

> |The list of IP addresses that must be denies access to the API. Add one or several IP addresses separated by a comma, or define a range. For more information, see <<specifying_ip_address,Specifying IP Addresses to Blacklist in the Policy>>.

Reviewer 2: Separated by comma? The UI does not allow it. You may define one IP or IP range at a time, as many times you need it, while configuring the policy .

Author: Done! Thanks!
